### PR TITLE
Enhance end to end test for Cardano hard forks

### PR DESCRIPTION
**Pull request by jpraynaud**
## Content
This PR includes improvements of the end to end test:
- Enhance wait strategy when writing era markers on chain in the `devnet`
- Fix transaction sent to the Cardano CLI when Cardano node is in **Conway** era
- Enhance the test cases of the CI:
  - Test multiple versions of the Cardano node: `8.7.3` and legacy `8.1.2`
  - Test hard fork transitions to **Conway** era at:
    - startup
    - few epochs after the startup
    - never

## Pre-submit checklist

- Branch
  - [x] Crates versions are updated (if relevant)
  - [x] Commit sequence broadly makes sense
  - [x] Key commits have useful messages
- PR
  - [x] No clippy warnings in the CI
  - [x] Self-reviewed the diff
  - [x] Useful pull request description
  - [x] Reviewer requested

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -262,22 +262,33 @@ jobs:
...
+        cardano_node_version: ["8.7.3"]
+        hard_fork_latest_era_at_epoch: [0,10,100]
+        run_id: ["#1","#2","#3"]
         extra_args: [""]
 
         include:
-          - era: ${{ fromJSON(needs.build-ubuntu-X64.outputs.eras)[0] }}
-            run_id: 3
-            hard_fork_latest_era_at_epoch: 1000
+          # Include a hard fork test for `8.1.2` which is not concerned`
```
Comment (dlachaumepalo) [inline .github/workflows/ci.yml:279]: Just the trailing character `
```suggestion
          # Include a hard fork test for `8.1.2` which is not concerned
```

**Review by dlachaumepalo — APPROVED** [review]
LGTM 🚀

File: .github/workflows/ci.yml (hunk 1 above)
Comment (jpraynaud) [reply .github/workflows/ci.yml]: Done :heavy_check_mark:

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -262,22 +262,33 @@ jobs:
...
       fail-fast: false
       matrix:
+        mode: ["std"]
         era: ${{ fromJSON(needs.build-ubuntu-X64.outputs.eras) }}
-        run_id: [1,2,3]
-        # Deactivate Conway era which is not currently working with the Pallas Chain Observer yet, value should be 5-10
-        hard_fork_latest_era_at_epoch: [1000] 
+        cardano_node_version: ["8.7.3"]
+        hard_fork_latest_era_at_epoch: [0,10,100]
+        run_id: ["#1","#2","#3"]
```
[the remark this answers is not in the document]
Comment (jpraynaud) [reply .github/workflows/ci.yml:275]: Done :heavy_check_mark:

File: mithril-test-lab/mithril-devnet/mkfiles/mkfiles-mithril.sh [hunk 3]
```diff
@@ -23,14 +23,31 @@ set -e
...
 
 cat >> era-mithril.sh <<EOF
+    # Get current Cardano era
+    CURRENT_CARDANO_ERA=\$(CARDANO_NODE_SOCKET_PATH=node-pool${N}/ipc/node.sock ./cardano-cli query tip \\
+        --testnet-magic ${NETWORK_MAGIC} \\
+        | jq  -r '.era |= ascii_downcase | .era')
+    echo ">>>> Current Cardano Era: \${CURRENT_CARDANO_ERA}"
+
+    # Fix: era related command is not (well) supported in Cardano node version '8.1.2'
+    if [ "${CARDANO_NODE_VERSION}" = "8.1.2" ]; then
```
[the remark this answers is not in the document]
Comment (jpraynaud) [reply mithril-test-lab/mithril-devnet/mkfiles/mkfiles-mithril.sh:33]: I agree 100%. 
We will be able to remove it once the `8.1.2` is deprecated as newer version embed era specific commands :+1:

File: mithril-test-lab/mithril-devnet/mkfiles/mkfiles-mithril.sh [hunk 4]
```diff
@@ -39,27 +56,52 @@ cat >> era-mithril.sh <<EOF
...
 
     ## Submit the transaction
-    CARDANO_NODE_SOCKET_PATH=node-pool${N}/ipc/node.sock ./cardano-cli transaction submit \\
+    CARDANO_NODE_SOCKET_PATH=node-pool${N}/ipc/node.sock ./cardano-cli \${CURRENT_CARDANO_ERA} transaction submit \\
         --tx-file node-pool${N}/tx/tx${N}-era-funds.tx \\
         --testnet-magic ${NETWORK_MAGIC}
 
-    ## Wait for the transaction to be confirmed
-    sleep 1
+    ## Wait at least for 2 blocks so that the transaction is confirmed
```
[the remark this answers is not in the document]
Comment (jpraynaud) [reply mithril-test-lab/mithril-devnet/mkfiles/mkfiles-mithril.sh:101]: Done :heavy_check_mark:

**Review by Alenar — APPROVED** [review]
:rocket:

**Pull request merged**